src/durable-cache/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+                            }
+                        }
+                    }
+                    ListenEvent::Progress(x) => {
+                        self.local_progress =
+                            x.into_option().expect("cache shard should not be closed")
+                    }
+                }
+            }
+        }
+        progress
+    }
+
+    /// Get and return the value associated with `key` if it exists, without syncing with the
+    /// durable store.
+    pub fn get_local(&self, key: &C::Key) -> Option<&C::Val> {
+        self.local.get(key)
+    }
+
+    /// Get and return the value associated with `key`, syncing with the durable store if
+    /// necessary. If `key` does not exist, then a value is computed via `val_fn` and durably
+    /// stored in the cache.
+    pub async fn get(&mut self, key: &C::Key, val_fn: impl FnOnce() -> C::Val) -> &C::Val {
+        // Fast-path if it's already in the local cache.
+        // N.B. This pattern of calling `contains_key` followed by `expect` is required to appease
+        // the borrow checker.
+        if self.local.contains_key(key) {
+            return self.local.get(key).expect("checked above");
+        }
+
+        // Reduce wasted work by ensuring we're caught up to at least the
+        // pubsub-updated shared_upper, and then trying again.
+        self.sync_to(self.write.shared_upper().into_option()).await;
+        if self.local.contains_key(key) {
+            return self.local.get(key).expect("checked above");
+        }
+
+        // Okay compute it and write it durably to the cache.
+        let val = val_fn();
+        let mut expected_upper = self.local_progress;
+        loop {
+            let update = ((C::encode_key(key), C::encode_val(&val)), expected_upper, 1);
+            let new_upper = expected_upper + 1;
+            let ret = self
+                .write
+                .compare_and_append(
+                    &[update],
+                    Antichain::from_elem(expected_upper),
+                    Antichain::from_elem(new_upper),
+                )
+                .await
+                .expect("usage should be valid");
+            match ret {
+                Ok(()) => {
+                    self.sync_to(Some(new_upper)).await;
+                    return self.get_local(key).expect("just inserted");
+                }
+                Err(err) => {
+                    expected_upper = self.sync_to(err.current.into_option()).await;
+                    if self.local.contains_key(key) {
+                        return self.local.get(key).expect("checked above");
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+
+    /// Durably set `key` to `value`. A `value` of `None` deletes the entry from the cache.
+    pub async fn set(&mut self, key: &C::Key, value: Option<&C::Val>) {
+        self.set_many(vec![(key, value)]).await
```

Comment:
An alternative here would be to re-implement the entire function without allocating a new `vec` for the argument.